### PR TITLE
Add credential controller logic

### PR DESCRIPTION
**Pull request by vitorarins**

### CODE REVIEW

**Comment by JorritSalverda** [comment]
It creates one configmap per CRD type, right?

**Comment by vitorarins** [comment]
> It creates one configmap per CRD type, right?

Yes. So this is only the logic for the Credentials for now. How it works: it creates one configmap named `estafette-external-credentials` when the first `Credential` CRD is deployed to the cluster.
Then for a new `Credential` it will only update the `estafette-external-credentials` configmap to add this new credential.

**Comment by JorritSalverda** [comment]
And if we run more than 1 Estafette server per Kubernetes cluster in separate namespaces it creates this configmap for each namespace?

File: controllers/credential_controller.go [hunk 1]
```diff
@@ -2,34 +2,268 @@ package controllers
...
 }
 
 // +kubebuilder:rbac:groups=ci.estafette.io,resources=credentials,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=ci.estafette.io,resources=credentials/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;create;update;patch;delete
 
 func (r *CredentialReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("credential", req.NamespacedName)
+	ctx := context.Background()
```
Comment (JorritSalverda) [inline controllers/credential_controller.go:37]: Doesn't the incoming `req` argument have a context we can reuse?

File: controllers/credential_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+					Namespace: "default",
+				},
+			}
+
+			_ = k8sClient.Delete(context.Background(), cred)
+		}
+	})
+
+	Context("One crededential", func() {
+		It("Should handle credential correctly", func() {
```
Comment (JorritSalverda) [inline controllers/credential_controller_test.go]: What is handling it correctly? Perhaps be a bit more explicit in saying what is correct behaviour.

File: controllers/credential_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      key.Name,
+					Namespace: key.Namespace,
+				},
+				Spec:   spec,
+				Status: civ1.CredentialStatus{},
+			}
+
+			By("Creating the credential successfully")
+			Expect(k8sClient.Create(context.Background(), toCreate)).Should(Succeed())
```
Comment (JorritSalverda) [inline controllers/credential_controller_test.go]: Can you a) put an `// act` comment above the line that executes the code under test and b) split the act and the assertion if that's possible? Meaning that if `k8sClient.Create` returns a result, assign it to a variable and then check whether it matches the expected value on a separate line? Now it combines those 2 things making it harder to read.

File: controllers/credential_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+			}
+			expectedYaml, err := yaml.Marshal(expectedCredentialsData)
+			Expect(err).To(BeNil())
+
+			Eventually(func() string {
+				_ = k8sClient.Get(context.Background(), configMapKey, fetchedConfigMap)
+				return fetchedConfigMap.Data["credentials-config.yaml"]
+			}, timeout, interval).Should(Equal(string(expectedYaml)))
+
+			By("Deleting the credential")
```
Comment (JorritSalverda) [inline controllers/credential_controller_test.go:128]: It seems this single test is verifying multiple functional requirements. If possible please split them into separate more accurately described tests so we know which functional requirement is tested and if it breaks what's actually broken.

File: controllers/credential_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+			}, timeout, interval).Should(Succeed())
+
+			Eventually(func() error {
+				f := &civ1.Credential{}
+				return k8sClient.Get(context.Background(), key, f)
+			}, timeout, interval).ShouldNot(Succeed())
+		})
+	})
+
+	Context("Two crededentials", func() {
```
Comment (JorritSalverda) [inline controllers/credential_controller_test.go]: Typo

**Review by JorritSalverda — CHANGES_REQUESTED** [review]

File: controllers/credential_controller.go (hunk 1 above)
Comment (vitorarins) [reply controllers/credential_controller.go:37]: It doesn't. Is not an http request, is controller reconcile request: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Request

Plus, this comes from recommendation of controller implementation.

**Review by vitorarins — COMMENTED** [review]
Will work on the changes.

**Comment by vitorarins** [comment]
> And if we run more than 1 Estafette server per Kubernetes cluster in separate namespaces it creates this configmap for each namespace?

Yes. You can see it by line 43 of credential_controller.go:
`r.List(ctx, &childConfigMaps, client.InNamespace(req.Namespace), client.MatchingFields{configMapOwnerKey: "Credential"})`

and line 49 is checking only the namespaced credential:
`r.Get(ctx, req.NamespacedName, &credential)`

**Comment by vitorarins** [comment]
@JorritSalverda Thanks for the tips. I caught a bug while splitting the test for deletion.
To be honest I don't like this ginkgo very much. I will probably get rid of that in the future.
But the documentation given by kubebuilder book points to an example using this ginkgo. So it was easier to have some tests already.

**Comment by vitorarins** [comment]
Merging this for now. If something needs to be changed it can be done in another PR.

**Pull request merged**